### DISCUSSION TITLE
[Proposal]: Support Validation on Method Parameters via Attributes

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#6063] [Language Ideas]
### Summary
Validating method parameters is a tricky problem space.  Sometimes you want to throw an exception if a value is invalid and sometimes you want to simply interrogate (not throw) if a value is invalid.  This proposal allows developers to validate method parameters via attributes.

### Motivation
C# 11 was contemplating adding a null validation operator (!!).  While this would remove a significant amount of boiler plate, it was a specific-case of a more general problem.  Also, while it did provide a mechanism to throw exceptions, it did not provide a method to interrogate whether values would throw or not.

### Detailed Design
A new attribute, ```ValidationAttribute``` shall be introduced.  When this attribute is present on another custom attribute, the compiler shall expect a static method with the following signatures to exist on the custom attribute itself:
```
 public static Exception? TryValidate<T>(T value, [CallerArgumentExpression("value")] string? ArgumentName = default) ;
```
In the signature, ```T``` may be a generic type parameter or a concrete type.  If multiple ```TryValidate``` methods exist, overload resolution rules shall select the appropriate overload.

If a method contains any parameters who attributes are decorated with ```ValidationAttribute```, the compiler shall synthesize the following code prior to the method body:
```
if (<ATTRIBUTETYPE>.TryValidate(<PARAMETER>) is { } Error1) {
    throw Error1;
}
```

### Example
Given the following code:
```
//Attribute definition
[ValidationAttribute]
[AttributeUsage(AttributeTargets.Parameter)]
public class RequiresNotNullAttribute : Attribute {
    public static Exception? TryValidate<T>(T value, [CallerArgumentExpression("value")] string? ArgumentName = default) {
        var ret = default(Exception?);
        
        if (value == null) {
            ret = new ArgumentNullException(ArgumentName);
        }

        return ret;
   }
}

//Attribute Usage:
public class ExampleClass {
  public void ExampleMethod([RequiresNotNull] string P1) {

      //Method Body

   }
}

```
The following will be generated:
public class ExampleClass {
  public void ExampleMethod([RequiresNotNull] string P1) {

    //Auto-Generated code
    {
      if(RequiresNotNullAttribute.TryValidate(P1) is {} Error1){
        throw Error1;
      }
    }

    //Method Body

   }
}

### Other Information
1. This is designed to be fast and performant.  Using a static method will eliminate allocations and virtual/interface method lookups
2. This is inspired InterpolatedStringHandler.
3. This is source-generator friendly.  Everything happens at compile time.
4. This is reflection-friendly.  Interrogating method parameter attributes will provide all validators.

### Limitations and Alternatives
As described above, a static ```TryValidate``` method must exist on the attribute.  This will provide optimal performance but limits usage.  For example, the following definition would not be valid:
```
  public void ExampleMethod([RequiresRange(5,10)] int P1);
```

As a fallback, if no applicable static methods are found, the compiler could instantiate the attribute and call an instance method at the cost of an allocation and an instance method invocation.

**Comment by HaloFour** [comment]
I'd much rather a more general AOP solution than one specific to parameter validation.

This also must overcome the assertion from the language design team that custom attributes never (rarely?) affect the runtime semantics of the code written by the developer.

**Reply by RikkiGibson** [reply]
I sometimes feel we are too cute about "attributes don't affect semantics". Clearly attributes affect runtime semantics, in a real sense, of serialization, ASP.NET routing, and a whole litany of other things. I would at least like to understand better why that line was initially drawn, and why "what we have now", and what people already do, doesn't cross that line, so to speak.

**Reply by TonyValenti** [reply]
@RikkiGibson I wasn't aware that was ever a goal/perspective, but, especially with ```InterpolatedStringHandler``` it seems like the perspective has changed.

**Reply by HaloFour** [reply]
> Clearly attributes affect runtime semantics, in a real sense, of serialization, ASP.NET routing, and a whole litany of other things. I would at least like to understand better why that line was initially drawn, and why "what we have now", and what people already do, doesn't cross that line, so to speak.

Every time this has come up on Discord the distinction has always been whether the attributes change the meaning and behavior of the code that you've written within a method, and it sounds like there are very few cases of that.  Those other attributes affect how a framework wires up your code with external components, they can't interject logic except for where that framework has control.

But I agree, I think that to a lot of developers that is an unsatisfactory answer and most are used to using attributes to declaratively affect how their code is executed.  There is a lot of potential for AOP solutions based on attributes and/or source generators that would satisfy the desire for parameter validation as well as many other scenarios.

**Comment by TahirAhmadov** [comment]
This is interesting. How would you solve parameters for these attributes?
```cs
void Foo([GreaterThan(5)]int x) { ... }
```

**Reply by TonyValenti** [reply]
@TahirAhmadov - I intentionally didn't mention that scenario because allocations are out of favor at the moment, however, I would propose the following solution:

```
{
  if (new GreaterThan(5).TryValidate(<PARAMETER>) is { } Error1) {
    throw Error1;
  }
}
```
